### PR TITLE
Add generic agent component registry

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Changes:
- add `AgentComponentRegistry`, a generic registry for `Component` objects
- remove `ProtocolRegistry` and `ContractRegistry` since they are subsumed by the `AgentComponentRegistry`

## Fixes

Preliminary work to address #1291 and #1290 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/registries/base.py [hunk 1]
```diff
@@ -107,132 +107,107 @@ def teardown(self) -> None:
...
+        :param component_id: the contract id
+        :return: the component or None if the component is not registered
         """
-        self._protocols = {}  # type: Dict[ProtocolId, Protocol]
+        return self._components_by_type.get(component_id.component_type, {}).get(
+            component_id.public_id, None
+        )
 
-    def register(self, item_id: PublicId, protocol: Protocol) -> None:
+    def fetch_all(self) -> List[Component]:
```
Comment (DavidMinarsch) [inline aea/registries/base.py:191]: Should this not be on a component type basis; i.e. require the component type as arg?

Comment (DavidMinarsch) [reply aea/registries/base.py:191]: Ah I see, exists below

File: aea/registries/resources.py [hunk 2]
```diff
@@ -65,16 +69,14 @@ def __init__(self, directory: Optional[Union[str, os.PathLike]] = None):
         :param directory: the path to the directory which contains the resources
              (skills, connections and protocols)
         """
-        self._contract_registry = ContractRegistry()
-        self._protocol_registry = ProtocolRegistry()
+        self._component_registry = AgentComponentRegistry()
         self._handler_registry = HandlerRegistry()
         self._behaviour_registry = ComponentRegistry[Behaviour]()
         self._model_registry = ComponentRegistry[Model]()
         self._skills = dict()  # type: Dict[SkillId, Skill]
```
Comment (DavidMinarsch) [inline aea/registries/resources.py:76]: Still an odd split? I.e. Skills treated very differently

Comment (DavidMinarsch) [reply aea/registries/resources.py:76]: Ok for now though!

Comment (marcofavorito) [reply aea/registries/resources.py:76]: yes, incoming PR will fix that

**Pull request merged**